File: library/src/main/java/com/mobilesolutionworks/android/app/controller/WorksControllerLifecycleHook.java
package com.mobilesolutionworks.android.app.controller;

import android.content.res.Configuration;
import android.os.Bundle;
import android.util.SparseArray;

/**
 * Created by yunarta on 8/3/17.
 */

public class WorksControllerLifecycleHook {

    private final WorksControllerManager mManager;

    public WorksControllerLifecycleHook(WorksControllerManager manager) {
        mManager = manager;
    }

    /**
     * Hook host dispatch pause to controller.
     * <p>
     * This method must be called by host.
     */
    public void dispatchPause() {
        mManager.dispatchPause();
        SparseArray<WorksController> controllers = mManager.getControllers();
        int size = controllers.size();
        for (int i = 0; i < size; i++) {
            controllers.valueAt(i).onPaused();
        }
    }

    /**
     * Hook host dispatch resume to controller.
     * <p>
     * This method must be called by host.
     */
    public void dispatchResume() {
        mManager.dispatchResume();
        SparseArray<WorksController> controllers = mManager.getControllers();
        int size = controllers.size();
        for (int i = 0; i < size; i++) {
            controllers.valueAt(i).onResume();
        }
    }

    /**
     * Hook host dispatch destroy to controller.
     * <p>
     * This method must be called by host.
     */
    void dispatchDestroy() {
        SparseArray<WorksController> controllers = mManager.getControllers();
        int size = controllers.size();
        for (int i = 0; i < size; i++) {
            controllers.valueAt(i).onDestroy();
        }
        controllers.clear();
    }

    public void onConfigurationChanged(Configuration config) {
        SparseArray<WorksController> controllers = mManager.getControllers();
        int size = controllers.size();
        for (int i = 0; i < size; i++) {
            controllers.valueAt(i).onConfigurationChanged(config);
        }
    }

    /**
     * Hook host dispatch restore instance to controller.
     * <p>
     * This method must be called by host.
     */
    public void onRestoreInstanceState(Bundle state) {
        if (state != null) {
            SparseArray<WorksController> controllers = mManager.getControllers();
            int size = controllers.size();
            for (int i = 0; i < size; i++) {
                Bundle bundle = state.getParcelable(":worksController:" + controllers.keyAt(i));
                controllers.valueAt(i).onViewStateRestored(bundle);
            }
        }
    }

    /**
     * Hook host dispatch save instance state instance to controller.
     * <p>
     * This method must be called by host.
     */
    public void dispatchSaveInstanceState(Bundle state) {
        SparseArray<WorksController> controllers = mManager.getControllers();
        int size = controllers.size();
        for (int i = 0; i < size; i++) {
            WorksController info = controllers.valueAt(i);

            Bundle bundle = new Bundle();
            info.onViewStateRestored(bundle);

            state.putParcelable(":worksController:" + controllers.keyAt(i), bundle);
        }
    }
}

File: library/src/main/kotlin/com/mobilesolutionworks/android/app/controller/WorksControllerLifecycleHook.kt
package com.mobilesolutionworks.android.app.controller

import android.content.res.Configuration
import android.os.Bundle

/**
 * Exposed hook for host to pass their lifecycle.
 *
 *
 * Created by yunarta on 8/3/17.
 */
class WorksControllerLifecycleHook internal constructor(private val mManager: WorksControllerManager) {

    /**
     * Hook host dispatch pause to controller.
     *
     *
     * This method must be called by host.
     */
    fun dispatchPause() {
        mManager.dispatchPause()
        val controllers = mManager.controllers
        val size = controllers.size()
        for (i in 0..size - 1) {
            controllers.valueAt(i).dispatchOnPaused()
        }
    }

    /**
     * Hook host dispatch resume to controller.
     *
     *
     * This method must be called by host.
     */
    fun dispatchResume() {
        mManager.dispatchResume()
        val controllers = mManager.controllers
        val size = controllers.size()
        for (i in 0..size - 1) {
            controllers.valueAt(i).dispatchOnResume()
        }
    }

    /**
     * Hook host dispatch destroy to controller.
     *
     *
     * This method must be called by host.
     */
    internal fun dispatchDestroy() {
        val controllers = mManager.controllers
        val size = controllers.size()
        for (i in 0..size - 1) {
            controllers.valueAt(i).onDestroy()
        }
        controllers.clear()
    }

    fun onConfigurationChanged(config: Configuration) {
        val controllers = mManager.controllers
        val size = controllers.size()
        for (i in 0..size - 1) {
            controllers.valueAt(i).onConfigurationChanged(config)
        }
    }

    /**
     * Hook host dispatch restore instance to controller.
     *
     *
     * This method must be called by host.
     */
    fun onRestoreInstanceState(state: Bundle?) {
        if (state != null) {
            val controllers = mManager.controllers
            val size = controllers.size()
            for (i in 0..size - 1) {
                val bundle = state.getParcelable<Bundle>(":worksController:" + controllers.keyAt(i))
                if (bundle != null) {
                    controllers.valueAt(i).dispatchOnRestoreInstanceState(bundle)
                }
            }
        }
    }

    /**
     * Hook host dispatch save instance state instance to controller.
     *
     *
     * This method must be called by host.
     */
    fun dispatchSaveInstanceState(state: Bundle) {
        val controllers = mManager.controllers
        val size = controllers.size()
        for (i in 0..size - 1) {
            val controller = controllers.valueAt(i)

            val bundle = Bundle()
            controller.dispatchOnSaveInstanceState(bundle)

            state.putParcelable(":worksController:" + controllers.keyAt(i), bundle)
        }
    }
}